File: Algorithm_117_Populating Next Right Pointers in Each Node II.cpp
#include <iostream>
#include <queue>
using namespace std;
struct TreeLinkNode
{
    int val;
    TreeLinkNode *left, *right, *next;
    TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
};
class Solution
{
  public:
    void connect(TreeLinkNode *root)
    {
        if (!root) return;
        queue<TreeLinkNode *> q;
        q.push(root);
        TreeLinkNode *lastNode = root;
        while (!q.empty())
        {
            TreeLinkNode *tmp = q.front();
            q.pop();
            if (tmp->left) q.push(tmp->left);
            if (tmp->right) q.push(tmp->right);
            if (tmp == lastNode)
            {
                tmp->next = NULL;
                lastNode = q.back();
            }
            else tmp->next = q.front();
        }
    }
};

// another solution
class Solution
{
    //based on level order traversal
  public:
    void connect(TreeLinkNode *root)
    {

        TreeLinkNode *head = NULL; //head of the next level
        TreeLinkNode *prev = NULL; //the leading node on the next level
        TreeLinkNode *cur = root;  //current node of current level

        while (cur != NULL)
        {

            while (cur != NULL)
            { //iterate on the current level
                //left child
                if (cur->left != NULL)
                {
                    if (prev != NULL)
                    {
                        prev->next = cur->left;
                    }
                    else
                    {
                        head = cur->left;
                    }
                    prev = cur->left;
                }
                //right child
                if (cur->right != NULL)
                {
                    if (prev != NULL)
                    {
                        prev->next = cur->right;
                    }
                    else
                    {
                        head = cur->right;
                    }
                    prev = cur->right;
                }
                //move to next node
                cur = cur->next;
            }

            //move to next level
            cur = head;
            head = NULL;
            prev = NULL;
        }
    }
}

File: Algorithm_117_Populating Next Right Pointers in Each Node II.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

// 思路：层序遍历，遍历的当前层的所有节点的next都已经设置完毕，所以可以直接用cur = cur.next来对这一层进行遍历
// 在遍历这一层的同时，将这层的下一层的所有节点连接起来，用nextLevelPre来记录下一层的上一个节点

class Solution {
    public Node connect(Node root) {
        Node cur = root, dummy = new Node(0), nextLevelPre = dummy;
        while (cur != null) {
            if (cur.left != null) {
                nextLevelPre.next = cur.left;
                nextLevelPre = nextLevelPre.next;
            }
            if (cur.right != null) {
                nextLevelPre.next = cur.right;
                nextLevelPre = nextLevelPre.next;
            }
            if (cur.next != null) { // still has node in this level
                cur = cur.next;
            } else { // switch to the next level
                cur = dummy.next;
                dummy = new Node(0);
                nextLevelPre = dummy;
            }
        }
        return root;
    }
}